File: src/Beta.Platform/Processors/RP6502/Interrupt.cs
using Beta.Platform.Exceptions;

namespace Beta.Platform.Processors.RP6502
{
    public static class Interrupt
    {
        const ushort nmiVector = 0xfffa;
        const ushort resVector = 0xfffc;
        const ushort irqVector = 0xfffe;

        public static void irq(InterruptState e, int value)
        {
            e.irq = value; // level sensitive
        }

        public static void nmi(InterruptState e, int value)
        {
            if (e.nmi_latch < value) // edge sensitive (0 -> 1)
            {
                e.nmi = 1;
            }

            e.nmi_latch = value;
        }

        public static void poll(InterruptState e, int i)
        {
            e.int_available = e.res | e.nmi | (e.irq & ~i);
        }

        public static ushort getVector(InterruptState e)
        {
            if (e.res == 1) { e.res = 0; return resVector; }
            if (e.nmi == 1) { e.nmi = 0; return nmiVector; }
            if (e.irq == 1) { e.irq = 0; return irqVector; }

            throw new CompilerPleasingException();
        }
    }
}

File: src/Beta.R6502/Interrupt.fs
namespace Beta.R6502


module Interrupt =

  let init =
    { int_available = 1uy;
      irq = 0uy;
      nmi = 0uy;
      nmi_latch = 0uy;
      res = 1uy;
    }


  let get_vector e =
    if e.res = 1uy then
      ({ e with res = 0uy }, 0xfffcus)
    elif e.nmi = 1uy then
      ({ e with nmi = 0uy }, 0xfffeus)
    else
      ({ e with irq = 0uy }, 0xfffaus)


  let irq e value =
    { e with
        irq = value;
    }


  let nmi e value =
    let edge =
      if e.nmi_latch < value
      then 1uy
      else 0uy
    in
      { e with
          nmi = e.nmi ||| edge
          nmi_latch = value
      }


  let poll e i =
    { e with
        int_available = e.res ||| e.nmi ||| (e.irq &&& ~~~i);
    }
